File: Medium/740.删除并获得点数.cpp
/*
 * @Author: Boredcui
 * @Date: 2022-03-03 20:38:17
 * @LastEditors: Boredcui
 * @LastEditTime: 2022-03-03 20:46:50
 * @FilePath: \LeetCode\740.删除并获得点数.cpp
 * @Description:
 *
 * Copyright (c) 2022 by boredcui, All Rights Reserved.
 */
/*
 * @lc app=leetcode.cn id=740 lang=cpp
 *
 * [740] 删除并获得点数
 */

// @lc code=start
class Solution
{
private:
    int rob(vector<int> &nums)
    {
        int size = nums.size();
        int first = nums[0], second = max(nums[0], nums[1]);
        for (int i = 2; i < size; ++i)
        {
            int temp = second;
            second = max(first + nums[i], second);
            first = temp;
        }
        return second;
    }

public:
    int deleteAndEarn(vector<int> &nums)
    {
        int maxVal = 0;
        for (int val : nums)
        {
            maxVal = max(val, maxVal);
        }
        vector<int> sum(maxVal + 1);
        for (int val : nums)
        {
            sum[val] += val;
        }
        return rob(sum);
    }
};
// @lc code=end

File: Medium/740.删除并获得点数.py
'''
Author: Boredcui
Date: 2022-03-03 20:07:15
LastEditors: Boredcui
LastEditTime: 2022-03-03 20:35:59
FilePath: \LeetCode\740.删除并获得点数.py
Description: 

Copyright (c) 2022 by boredcui, All Rights Reserved. 
'''
#
# @lc app=leetcode.cn id=740 lang=python3
#
# [740] 删除并获得点数
#

# @lc code=start


class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        total = [0]*(max(nums)+1)
        for val in nums:
            total[val] += val

        def rob(nums: List[int]) -> int:
            n = len(nums)
            first, second = nums[0], max(nums[0], nums[1])
            for i in range(2, n):
                first, second = second, max(first+nums[i], second)
            return second

        return rob(total)
# @lc code=end
